tiledb/sm/crypto/test/unit_crypto.cc

```diff
@@ -493,50 +517,113 @@ TEST_CASE("Crypto: Test AES-256-GCM", "[crypto][aes]") {
-    char shastring[65];
-    for (uint64_t i = 0; i < output_buffer.alloced_size(); ++i) {
-      sprintf(&shastring[i * 2], "%02x", (unsigned int)digest[i]);
-    }
+static std::string to_hex(const uint8_t* data, uint64_t length) {
+  std::stringstream ss;
+  for (uint64_t i = 0; i < length; i++) {
+    ss << std::hex << std::setw(2) << std::setfill('0') << (int)data[i];
+  }
+  return ss.str();
+}
+
+// Test that the given input has the expected hash value in
+// hex. The function is generic over the hash and the input type.
+template <class Hash>
+static void test_hash(
+    const uint8_t* input, uint64_t length, const std::string& expected_hash) {
+  REQUIRE(expected_hash.length() % 2 == 0);
 
-    CHECK(
-        memcmp(
-            expected_checksum.data(), shastring, expected_checksum.length()) ==
-        0);
+  Buffer hash_buf(Hash::digest_bytes);
+  CHECK((Hash::hash(input, length, &hash_buf)).ok());
+
+  // Compare the strings for a better error message in case of failure.
+  CHECK(
+      to_hex((uint8_t*)hash_buf.data(), hash_buf.alloced_size()) ==
+      expected_hash);
+}
+
+struct MD5Hash {
+  static const int digest_bytes = Crypto::MD5_DIGEST_BYTES;
+  static Status hash(const uint8_t* input, uint64_t length, Buffer* output) {
+    return Crypto::md5(input, length, output);
+  }
+};
+
+TEST_CASE("Crypto: Test MD5", "[crypto][md5]") {
+  auto test_md5 = [](const std::string input, const std::string expected_hash) {
+    test_hash<MD5Hash>((uint8_t*)(input.data()), input.length(), expected_hash);
+  };
+  // Values taken from section A.5 in https://www.ietf.org/rfc/rfc1321.txt
+  test_md5("", "d41d8cd98f00b204e9800998ecf8427e");
+  test_md5("a", "0cc175b9c0f1b6a831c399e269772661");
+  test_md5("abc", "900150983cd24fb0d6963f7d28e17f72");
+  test_md5("message digest", "f96b697d7cb7938d525a2f31aaf161d0");
+  test_md5("abcdefghijklmnopqrstuvwxyz", "c3fcd3d76192e4007dfb496cca67e13b");
+  test_md5(
+      "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789",
+      "d174ab98d277d9f5a5611c2c9f419d9f");
+  test_md5(
+      "1234567890123456789012345678901234567890123456789012345678901234567890"
+      "1234567890",
+      "57edf4a22be3c955ac49da2e2107b67a");
+}
+
+struct SHA256Hash {
+  static const int digest_bytes = Crypto::SHA256_DIGEST_BYTES;
+  static Status hash(const uint8_t* input, uint64_t length, Buffer* output) {
+    return Crypto::sha256(input, length, output);
   }
+};
+
+TEST_CASE("Crypto: Test SHA256", "[crypto][sha256]") {
+  auto test_sha256 = [](const std::string& input,
+                        const std::string& expected_hash) {
+    auto data = from_hex(input);
+    test_hash<SHA256Hash>(data.data(), data.size(), expected_hash);
+  };
+  // Values taken from SHA256ShortMsg.rsp, from "SHA Test Vectors for Hashing
+  // Bit-Oriented Messages", found at
+  // https://csrc.nist.gov/Projects/cryptographic-algorithm-validation-program/Secure-Hashing
+  // Len = 0
+  test_sha256(
```

Comment:
Here and above, it will be clearer to put test vectors into an array and loop through the array.

Reply:
Done.